---
title: "Differential expression analysis `r params$sample_name`"
author: 'User ID: `r Sys.getenv("USER")`'
date: "`r Sys.Date()`"
output:
  html_document:
    fig_height: 7
    fig_width: 9
    keep_md: yes
    toc: yes
    df_print: paged
params:
  out_path: "/gpfs/data/aifantislab/home/nadorb01/projects/ad_ped_AML_v5/results/"
  data_path: "/gpfs/data/aifantislab/home/nadorb01/projects/ad_ped_AML_v5/results/R-Figure3-B_cells_no_cc_reg/seurat_obj_B_fin.rds.rds"
  sample_name: "Atypical_B"
---

```{r setup, include=FALSE}
attach(params) 
library(Seurat)
library(future)
library(dplyr)
library(ggplot2)
library(tibble)
library(stringr)
library(MAST)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(clusterProfiler)
source("/gpfs/data/aifantislab/home/nadorb01/common_scripts/DE_helper.R")


knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.path = file.path(out_path, paste0("07-DE-", sample_name, "/")))



# enable parallelization
plan("multiprocess", workers = 4) # 10 cpus
options(future.globals.maxSize = 59600 * 1024^3) # for 64 Gb RAM, 119200 * 1024^3 for 128GB

# set out_path and create dir 
out_path1 <- out_path
out_path <- paste0(out_path, "07-DE-", sample_name, "/")
if (!dir.exists(out_path)) {dir.create(out_path)}

```


```{r read_data}
seurat_obj <- readRDS(data_path)
seurat_obj@meta.data$aml <- ifelse(str_detect(seurat_obj@meta.data$samples, "Control"), "Control", "AML")
```

```{r DE_Atypical_B}
# subset to atypical B cells
seurat_obj <-subset(seurat_obj, cells=rownames(seurat_obj@meta.data)[which(seurat_obj@meta.data$Cell_type_identity=="Atypical memory")])
```



```{r ME_maliget_XY_genes}
## rm genes from X/Y chromosomes ###
ensembl = biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl", host = "jul2018.archive.ensembl.org")
genes_tbl = biomaRt::getBM(attributes =
    c("ensembl_gene_id", "hgnc_symbol", "chromosome_name", "start_position", "end_position"),
    filters = "hgnc_symbol", values = rownames(seurat_obj), mart = ensembl)

genes_tbl = genes_tbl %>%
  filter(chromosome_name %in% c( "X", "Y"))
```


```{r AML_vs_control}
DefaultAssay(seurat_obj) <- "RNA_soupx"
print(table(seurat_obj@meta.data[,"aml"]))

all.genes <- rownames(seurat_obj)[which(rowSums(as.matrix(seurat_obj@assays$RNA@counts))>0)]
seurat_obj <- seurat_obj %>% 
                            NormalizeData(normalization.method = "LogNormalize", scale.factor = 10000) %>%
                            ScaleData(features = all.genes)

Idents(seurat_obj) <- "aml"

DE_genes <- FindMarkers(seurat_obj, ident.1 = "AML", min.pct = 0.1,only.pos = FALSE, test.use = "MAST")
DE_genes <- subset(DE_genes, !(rownames(DE_genes) %in% genes_tbl$hgnc_symbol))

write.csv(DE_genes, paste0(out_path, "DE_genes_Atypical_B_cells_MAST.csv"))

DE_genes_volcano <- FindMarkers(seurat_obj, ident.1 = "AML", min.pct = 0.1, logfc.threshold=0, test.use = "MAST")
DE_genes_volcano <- subset(DE_genes_volcano, !(rownames(DE_genes_volcano) %in% genes_tbl$hgnc_symbol))

top_up <- DE_genes_volcano %>% dplyr::filter(avg_logFC > 0.25) %>% dplyr::arrange(p_val_adj)
top_down <- DE_genes_volcano %>% dplyr::filter(avg_logFC < 0.25) %>% dplyr::arrange(avg_logFC) %>% dplyr::arrange(p_val_adj)

p <- EnhancedVolcano(DE_genes_volcano,
    lab = rownames(DE_genes_volcano),
    x = 'avg_logFC',
    y = 'p_val_adj',
    selectLab= c(rownames(top_up)[1:8],rownames(top_down)[1:8]),
    title = paste0('Atypical B cells - AML vs Control'),
    drawConnectors=TRUE,
    col=c("black","black","black","red3"),
    ylab = bquote(~-Log[10]~ 'P adj'),
    subtitle = NULL,
    pCutoff = 5e-2,
    FCcutoff=0.25)

print(p)
ggsave(paste0("Volcano_plot_","_Atypical_B_cells_AML_vs_Control_MAST.pdf"), plot=p, path=out_path, device="pdf", height=7, width=10)

# GO analysis
GO_terms <- run_GO(s_obj=seurat_obj, DE_genes=DE_genes, s_name="Atypical_B_cells")
GO_comparison_down_MF <- GO_terms$down_genes_MF$ID
GO_comparison_up_MF <- GO_terms$up_genes_MF$ID


saveRDS(DE_genes_volcano, paste0(out_path, "DE_genes_volcano_list_Atypical_B_cells.rds"))
saveRDS(seurat_obj,  paste0(out_path, "seurat_obj_Atypical_B_cells.rds"))
```
